src/wp-includes/webfonts-api/class-wp-webfonts-controller.php

```diff
@@ -0,0 +1,198 @@
+
+		// Add resources hints.
+		add_filter( 'wp_resource_hints', array( $this, 'get_resource_hints' ), 10, 2 );
+	}
+
+	/**
+	 * Generate and enqueue webfonts styles.
+	 *
+	 * @since 5.9.0
+	 */
+	public function generate_and_enqueue_styles() {
+		// Generate the styles.
+		$styles = $this->generate_styles();
+
+		// Enqueue the stylesheet.
+		wp_register_style( $this->stylesheet_handle, '' );
+		wp_enqueue_style( $this->stylesheet_handle );
+
+		// Add the styles to the stylesheet.
+		wp_add_inline_style( $this->stylesheet_handle, $styles );
+	}
+
+	/**
+	 * Generate and enqueue editor styles.
+	 *
+	 * @since 5.9.0
+	 */
+	public function generate_and_enqueue_editor_styles() {
+		wp_add_inline_style( 'wp-block-library', $this->generate_styles() );
+	}
+
+	/**
+	 * Generate styles for webfonts.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @return string $styles Generated styles.
+	 */
+	private function generate_styles() {
+		$styles    = '';
+		$providers = $this->get_providers()->get_all_registered();
+		foreach ( $providers as $provider_id => $provider ) {
+			$registered_webfonts = $this->webfonts_registry->get_by_provider( $provider_id );
+
+			if ( empty( $registered_webfonts ) ) {
+				continue;
+			}
+
+			$provider->set_webfonts( $registered_webfonts );
+			$styles .= $provider->get_css();
+		}
+		return $styles;
+	}
+
+	/**
+	 * Get the webfonts registry.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @return WP_Webfonts_Registry
+	 */
+	public function get_webfonts_registry() {
+		return $this->webfonts_registry;
+	}
+
+	/**
+	 * Get the providers registry.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @return WP_Webfonts_Provider_Registry
+	 */
+	public function get_providers() {
+		return $this->providers;
+	}
+
+	/**
+	 * Get the resource hints.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param array  $urls {
+	 *     Array of resources and their attributes, or URLs to print for resource hints.
+	 *
+	 *     @type array|string ...$0 {
+	 *         Array of resource attributes, or a URL string.
+	 *
+	 *         @type string $href        URL to include in resource hints. Required.
+	 *         @type string $as          How the browser should treat the resource
+	 *                                   (`script`, `style`, `image`, `document`, etc).
+	 *         @type string $crossorigin Indicates the CORS policy of the specified resource.
+	 *         @type float  $pr          Expected probability that the resource hint will be used.
+	 *         @type string $type        Type of the resource (`text/html`, `text/css`, etc).
+	 *     }
+	 * }
+	 * @param string $relation_type The relation type the URLs are printed for,
+	 *                              e.g. 'preconnect' or 'prerender'.
+	 *
+	 * @return array URLs to print for resource hints.
+	 */
+	public function get_resource_hints( $urls, $relation_type ) {
+		$providers = $this->get_providers()->get_all_registered();
+		foreach ( $providers as $provider_id => $provider ) {
```

Comment:
`$provider_id` is not used here.

```suggestion
		foreach ( $providers as $provider ) {
```